cpp/src/fil/internal.cuh

```diff
@@ -266,6 +295,122 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+__host__ __device__ __forceinline__ bool fetch_bit(const uint8_t* array, int bit)
+{
+  const int bits_per_byte = 8;
+  return array[bit / bits_per_byte] & (1 << bit % bits_per_byte);
+}
+
+struct cat_feature_counters {
+  int max_matching = -1, n_nodes = 0;
+  void reduce_with(cat_feature_counters b)
+  {
+    max_matching = std::max(max_matching, b.max_matching);
+    n_nodes += b.n_nodes;
+  }
+};
+struct categorical_sets {
+  // arrays from each node ID are concatenated first, then from all categories
+  const uint8_t* bits = nullptr;
+  // largest matching category in the model, per feature ID
+  const int* max_matching = nullptr;
+  size_t bits_size = 0, max_matching_size = 0;
+
+  __host__ __device__ __forceinline__ bool branch_can_be_categorical() const
+  {
+    // If this is constructed from cat_sets_owner, will return true
+    // default-initialized will return false
+    // Defining edge case: there are categorical nodes, but all have max_matching == -1
+    // (all categorical nodes are empty). node.thresh() would have returned 0.0f
+    // and the branch condition wouldn't have always been false (i.e branched left).
+    // Alternatively, we could have converted all empty categorical nodes to
+    // NAN-threshold numerical nodes.
+    return max_matching != nullptr;
+  }
+
+  // set count is due to tree_idx + node_within_tree_idx are both ints, hence uint32_t result
+  template <bool branch_can_be_categorical, typename node_t>
+  __host__ __device__ __forceinline__ int category_matches(node_t node, int category) const
+  {
+    // standard boolean packing. This layout has better ILP
+    // node.set() is global across feature IDs and is an offset (as opposed
+    // to set number). If we run out of uint32_t and we have hundreds of
+    // features with similar categorical feature count, we may consider
+    // storing node ID within nodes with same feature ID and look up
+    // {.max_matching, .first_node_offset} = ...[feature_id]
+    return category <= max_matching[node.fid()] && fetch_bit(bits + node.set(), category);
+  }
+  static int sizeof_mask_from_max_matching(int max_matching)
+  {
+    return raft::ceildiv(max_matching + 1, 8);
```

Comment:
Where does 8 come from? If it's `BITS_PER_BYTE`, then it's better to define it as a constant in the FIL namespace.